sdks/python/apache_beam/runners/interactive/utils.py

```diff
@@ -0,0 +1,112 @@
+# limitations under the License.
+#
+
+"""Utilities to be used in  Interactive Beam.
+"""
+
+from __future__ import absolute_import
+
+import pandas as pd
+
+from apache_beam.typehints import typehints as th
+from apache_beam.utils.windowed_value import WindowedValue
+
+COLUMN_PREFIX = 'el'
+
+
+def parse_row_(el, element_type, depth):
+  elements = []
+  columns = []
+
+  # Recurse if there are a known length of columns to parse into.
+  if isinstance(element_type, (th.TupleHint.TupleConstraint)):
+    for index, t in enumerate(element_type._inner_types()):
+      underlying_columns, underlying_elements = parse_row_(el[index], t,
+                                                           depth + 1)
+      column = '[{}]'.format(index)
+      if underlying_columns:
+        columns += [column + c for c in underlying_columns]
+      else:
+        columns += [column]
+      elements += underlying_elements
+
+  # Don't make new columns for variable length types.
+  elif isinstance(
+      element_type,
+      (th.ListHint.ListConstraint, th.TupleHint.TupleSequenceConstraint)):
+    elements = [pd.array(el)]
+
+  # For any other types, try to parse as a namedtuple, otherwise pass element
+  # through.
+  else:
+    fields = getattr(el, '_fields', None)
+    if fields:
+      columns = list(fields)
+      if depth > 0:
+        columns = ['[{}]'.format(f) for f in fields]
+      elements = [el._asdict()[f] for f in fields]
+    else:
+      elements = [el]
+  return columns, elements
+
+
+def parse_row(el, element_type, include_window_info=True, prefix=COLUMN_PREFIX):
+  # Reify the WindowedValue data to the Dataframe if asked.
+  windowed = None
+  if isinstance(el, WindowedValue):
+    if include_window_info:
+      windowed = el
+    el = el.value
+
+  # Parse the elements with the given type.
+  columns, elements = parse_row_(el, element_type, 0)
+
+  # If there are no columns returned, there is only a single column of a
+  # primitive data type.
+  if not columns:
+    columns = ['']
+
+  # Add the prefix to the columns that have an index.
+  for i in range(len(columns)):
+    if columns[i] == '' or columns[i][0] == '[':
+      columns[i] = prefix + columns[i]
+
+  # Reify the windowed columns and do a best-effort casting into Pandas DTypes.
+  if windowed:
+    columns += ['event_time', 'windows', 'pane_info']
+    elements += [
+        windowed.timestamp.micros, windowed.windows, windowed.pane_info
+    ]
+  return columns, elements
+
+
+def pcoll_to_df(
+    elements, element_type, include_window_info=False, prefix=COLUMN_PREFIX):
+  """Parses the given elements into a Dataframe.
+
+  Each column name will be prefixed with `prefix` concatenated with the nested
+  index, e.g. for a Tuple[Tuple[int, str], int], the column names will be:
+  [prefix[0][0], prefix[0][1], prefix[0]]. This is subject to change.
+  """
+  rows = []
+  columns = []
+
+  for e in elements:
+    columns, row = parse_row(e, element_type, include_window_info, prefix)
```

Comment:
Would it be better to infer the schema once, then apply all rows, rather than re-parse it for every element? (Also, that way you could correctly detect/handle if the schema is not consistent, e.g. tuples of varying length.)

Reply:
Changed to inferring the schema once.